### ISSUE TITLE
Error in addForeignKeyConstraint when using overridden table names

### ISSUE THREAD

**Issue by HemanthD** [jhipster/generator-jhipster#3723]
##### **Overview of the issue**

Liquibase generated addForeignKeyConstraint seems to use the default entity-name-derived table name rather than the specified/ overridden tablename (using --table-name).

Consider 2 entities Criteria and Option with one-to-many relationship. Both entities are generated using --table-name option as pa_criteria and pa_option. The foreign key generated in pa_option has referencedTableName="criteria" (default table name of Criteria entity) instead of referencedTableName="pa_criteria" (specified table name).

Liquibase fails while creating the fk.

liquibase.exception.DatabaseException: Cannot add foreign key constraint [Failed SQL: ALTER TABLE atlas.pa_option ADD CONSTRAINT fk_option_criteria_id FOREIGN KEY (criteria_id) REFERENCES atlas.criteria (id)]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:301)

##### **Motivation for or Use Case** 

Table names can be overridden using --table-name. However, there seem to be an issue when creating references to such entities - requiring manual intervention.

##### **JHipster Version(s)** 

3.3.0

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder** 

[.yo-rc.json.txt](https://github.com/jhipster/generator-jhipster/files/317725/default.yo-rc.json.txt)

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory** 

[Criteria.json.txt](https://github.com/jhipster/generator-jhipster/files/317728/Criteria.json.txt)
[Option.json.txt](https://github.com/jhipster/generator-jhipster/files/317727/Option.json.txt)

##### **Browsers and Operating System** 

Chrome on Win 8.1 with MySQL 5.7

##### **Reproduce the error** 

    save attached Criteria.json and Option.json in .jhipster/
    yo jhipster:entity criteria --table-name pa_criteria --regenerate
    yo jhipster:entity option --table-name pa_option --regenerate
    check the xxxx_added_entity_constraints_Option.xml - it will have the incorrect referencedTableName="criteria"
    launch the app
    Liquibase will throw an error while processing the xxxx_added_entity_constraints_Option.xml
    
##### **Related issues** 

Did not find any similar cases.

##### **Suggest a Fix** 

The issue could be with the below piece of code

    <addForeignKeyConstraint baseColumnNames="<%= getColumnName(relationshipName) %>_id"
                             baseTableName="<%= entityTableName %>"
                             constraintName="<%= constraintName %>"
                             referencedColumnNames="id"
                             referencedTableName="<%
                             var otherEntityTable = getTableName(otherEntityName);
                             if (otherEntityTable == 'user') { %>jhi_user<% } else { %><%=otherEntityTable %><% } %>"/>

in generator-jhipster/generators/entity/templates/src/main/resources/config/liquibase/changelog/_added_entity_constraints.xml which is using _otherEntityName_ for _getTableName()_. Perhaps, it should use the *'entityTableName'* attribute from the _otherEntity.json_

**Comment by deepu105** [comment]
hmm thats becoming more complex thats why we didnt want to support custom
table names initially to vaoid making stuff more complex. Do you want to do
a PR to fix this?


Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
Yes, overriding table names is a source of issues... and I don't even understand what you gain from it.

**Comment by jdubois** [comment]
When using a custom table name, the issue is that when you do joins, the other entities cannot know your table name as it's not the default anymore....

3 solutions for this:

- Add a new question: problem is that, you can't know if the joined table has a specific table name or not, so you're going to need to ask this question all the time!! That means everybody is going to get that question: that's annoying everyone for a very minor use case.
- Read the other entity's `.json` config file, to get its table name -> I don't think we've already done that, and this looks complex
- Don't generate Liquibase code with Yeoman, but use the `mvn liquibase:diff` goal instead, as this one knows about the table name. We don't do it at the moment (generating the diffs ourselves) mostly because we want the generation to be a one-step process. It would be quite annoying to run `yo` and then `mvn` just aftewards.

So basically I can't find a good solution here. The 2nd option looks the best to me, and opens some interesting opportunities, but this looks complex for such a use case.

So I have 2 proposals:

- We remove the "table-name" option, as I don't even understand the need for this. If your entities don't have the names of your tables, it's just going to make things more complex, and you won't gain anything from it! Besides, if you want to change this manually after generation, it's not that hard: just modify the annotation, and run `liquibase:diff` again -> for me this option is even against our [Policy 2](http://jhipster.github.io/policies/)
- Do option 2 from above, reading the other entity's JSON configuration
